File: samples/ScanUI/ScanUI/ScanTableViewController.m
//
//  ScanTableViewController.m
//  ScanUI
//
//  Created by Takeshi Shoji on 8/22/16.
//  Copyright © 2016 Novars Inc. All rights reserved.
//

#import "ScanTableViewController.h"
@import MaBeeeSDK;

@interface ScanTableViewController ()
@property (nonatomic, strong) NSArray<MaBeeeDevice *> *devices;
@end

@implementation ScanTableViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc]
                                           initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                           target:self
                                           action:@selector(doneButtonPressed:)];
}

- (void)doneButtonPressed:(UIBarButtonItem *)doneButton {
  [self dismissViewControllerAnimated:YES completion:NULL];
}

- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];
  [MaBeeeApp.instance addObserver:self selector:@selector(receiveNotification:)];
  [MaBeeeApp.instance startScan:^(NSArray<MaBeeeDevice *> *devices) {
    self.devices = devices;
    [self.tableView reloadData];
  }];
}

- (void)viewWillDisappear:(BOOL)animated {
  [super viewWillDisappear:animated];
  [MaBeeeApp.instance removeObserver:self];
  [MaBeeeApp.instance stopScan];
}

- (void)receiveNotification:(NSNotification *)notification {
  [self.tableView reloadData];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.devices.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MaBeeeDeviceCell"];
  MaBeeeDevice *device = self.devices[indexPath.row];
  NSString *state = [self stateString:device];
  cell.textLabel.text = [NSString stringWithFormat:@"%@, %d, %@", device.name, device.rssi, state];
  return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:YES];
  MaBeeeDevice *device = self.devices[indexPath.row];
  switch (device.state) {
    case MaBeeeDeviceStateDisconnected:
      [MaBeeeApp.instance connect:device];
      break;
    case MaBeeeDeviceStateConnecting:
      [MaBeeeApp.instance disconnect:device];
      break;
    case MaBeeeDeviceStateConnected:
      [MaBeeeApp.instance disconnect:device];
      break;
  }
}

- (NSString *)stateString:(MaBeeeDevice *)device {
  NSString *stateString;
  switch (device.state) {
    case MaBeeeDeviceStateDisconnected:
      stateString = @"Disconnected";
      break;
    case MaBeeeDeviceStateConnecting:
      stateString = @"Connecting";
      break;
    case MaBeeeDeviceStateConnected:
      stateString = @"Connected";
      break;
  }
  return stateString;
}

@end

File: samples/ScanUISwift/ScanUISwift/ScanTableViewController.swift
//
//  ScanTableViewController.swift
//  ScanUISwift
//
//  Created by Takeshi Shoji on 8/22/16.
//  Copyright © 2016 Novars Inc. All rights reserved.
//

import UIKit
import MaBeeeSDK

class ScanTableViewController: UITableViewController {

  var maBeeeDevices:[MaBeeeDevice]?

  override func viewDidLoad() {
    super.viewDidLoad()
    self.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done,
                                                            target: self,
                                                            action: #selector(doneButtonPressed))
  }

  func doneButtonPressed() {
    self.dismiss(animated: true, completion: nil)
  }

  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    MaBeeeApp.instance().addObserver(self, selector: #selector(receiveNotification(_:)))
    MaBeeeApp.instance().startScan({(devices: [MaBeeeDevice]?) in
      print(devices ?? "nil")
      self.maBeeeDevices = devices
      self.tableView.reloadData()
    })
  }

  override func viewWillDisappear(_ animated: Bool) {
    super.viewDidAppear(animated)
    MaBeeeApp.instance().removeObserver(self)
    MaBeeeApp.instance().stopScan()
  }

  func receiveNotification(_ notification: Notification) {
    self.tableView.reloadData()
  }

  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    if let devices = maBeeeDevices {
      return devices.count;
    }
    return 0;
  }

  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "MaBeeeDeviceCell", for: indexPath)
    if let device = maBeeeDevices?[indexPath.row] {
      cell.textLabel?.text = String(format: "%@, %d, %@",
                                    device.name,
                                    device.rssi,
                                    self.stateString(device.state))
    }
    return cell;
  }

  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    tableView.deselectRow(at: indexPath, animated: true)
    if let device = maBeeeDevices?[indexPath.row] {
      switch (device.state) {
      case .disconnected:
        MaBeeeApp.instance().connect(device)
      case .connecting:
        MaBeeeApp.instance().disconnect(device)
      case .connected:
        MaBeeeApp.instance().disconnect(device)
      }
    }
  }


  func stateString(_ state: MaBeeeDeviceState) -> String {
    var stateString: String
    switch (state) {
    case .disconnected:
      stateString = "Disconnected"
    case .connecting:
      stateString = "Connecting"
    case .connected:
      stateString = "Connected"
    }
    return stateString
  }
}
